File: java/leetCode/twoPoint/Problem021.java
package leetCode.twoPoint;

/**
 * Merge two sorted linked lists and return it as a new list.
 * The new list should be made by splicing together the nodes of the first two lists.
 * <p>
 * Example:
 * Input: 1->2->4, 1->3->4
 * Output: 1->1->2->3->4->4
 * <p>
 * create by stephen on 2018/5/15
 */
public class Problem021 {

    private class ListNode {
        int val;
        ListNode next;

        ListNode(int x) {
            val = x;
        }
    }

    /**
     * 遍历 两个指针移动
     * 用一个dummy节点表示公共的head节点
     * O(n) time + O(a) space
     */
    @SuppressWarnings("all")
    public ListNode mergeTwoLists1(ListNode l1, ListNode l2) {
        ListNode dummy = new ListNode(Integer.MIN_VALUE);
        ListNode tail = dummy;

        while (l1 != null && l2 != null) {
            if (l1.val < l2.val) {
                tail.next = l1;
                l1 = l1.next;
            } else {
                tail.next = l2;
                l2 = l2.next;
            }
            tail = tail.next;
        }

        tail.next = l1 != null ? l1 : l2;
        return dummy.next;
    }

    /**
     * 递归解法
     */
    public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        if (l1 == null) return l2;
        if (l2 == null) return l1;

        if (l1.val < l2.val) {
            l1.next = mergeTwoLists(l1.next, l2);
            return l1;
        } else {
            l2.next = mergeTwoLists(l2.next, l1);
            return l2;
        }
    }

}

File: leetCode/linkedList/Problem021.cpp
#include "list.h"

ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
  ListNode dummy;
  ListNode* cur = &dummy;
  while (l1 != nullptr || l2 != nullptr) {
    if (l1 != nullptr && l2 != nullptr) {
      if (l1->val < l2->val) {
        cur->next = l1;
        l1 = l1->next;
      } else {
        cur->next = l2;
        l2 = l2->next;
      }
      cur = cur->next;
    } else if (l1 != nullptr) {
      cur->next = l1;
      l1 = l1->next;
      cur = cur->next;
    } else {
      cur->next = l2;
      l2 = l2->next;
      cur = cur->next;
    }
  }
  return dummy.next;
}